core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp

```diff
@@ -820,27 +820,83 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs < tmp_rhs;
   }
 
+  template <class T>
+  KOKKOS_FUNCTION friend std::enable_if_t<
+      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
+  operator<(floating_point_wrapper lhs, T rhs) {
+    return T(lhs) < rhs;
+  }
+
+  template <class T>
+  KOKKOS_FUNCTION friend std::enable_if_t<
+      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
+  operator<(T lhs, floating_point_wrapper rhs) {
+    return lhs < T(rhs);
+  }
```

Comment:
Wouldn't it be better to provide the overload only for `float` (and accept a possible narrowing conversion)? Types convertible to `float` should then still work (but don't with the current approach since both overloads would be feasible comparing, e.g., with an `int`).
 ```suggestion
  KOKKOS_FUNCTION friend bool
  operator<(floating_point_wrapper lhs, float rhs) {
    return T(lhs) < rhs;
  }
```

Reply:
Note that the `SYCL` implementation just casts everything to `float`: https://github.com/intel/llvm/blob/83aafddac12e9b2df8e1764d3fba0ffe7ef6c596/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L176-L198.

Reply:
> Wouldn't it be better to provide the overload only for `double`? Types convertible to `double` should then still work (but don't with the current approach since both overloads would be feasible comparing, e.g., with an `int`

I like the simplicity of this approach. I believe that for other mixed precision expressions involving `T1 op T2`, the compiler will implicitly up-cast either `T1` or `T2`. So, providing only the double overload would result in unexpected behavior by up-casting half-precision operands to double. That is my understanding.

Reply:
For a comparison with an `int`, https://godbolt.org/z/efnxf7e5n gives an error:
```
ambiguous overload for 'operator<' (operand types are 'Foo' and 'int')
```

If only providing the `double` or `float` overload we are getting a similar warning with GCC and an error with clang (if the conversion operator is not explicit).

If we use one templated version, everything works fine, and we can do whatever we want in the comparison operator. This does different casts depending on the types used, see https://cppinsights.io/s/2155d1dd.

If we don't provide the operator, implicit conversion kicks in which means we always cast this type to float and we are then doing comparisons based on the other type used, see https://cppinsights.io/s/1c62be16. It's not quite clear to me where this is failing at the moment.
If implicit conversion isn't sufficient so that we actually need to declare comparison operators, it seems most sensible to me to provide a templated overload that cast the half type to `float` and let the compiler deal with other casts required.

Reply:
With the current approach we are getting
```
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:903:52: error: invalid operands to binary expression ('int' and 'const Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>')
    actual_lhs(LT_S_H)   = static_cast<int>(h_lhs) < h_rhs;
                           ~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:377:7: note: in instantiation of member function 'Test::Functor_TestHalfOperators<Kokkos::View<double *, Kokkos::Experimental::SYCL>, Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>::operator()' requested here
      run_on_host(0);
      ^
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:980:50: note: in instantiation of member function 'Test::Functor_TestHalfOperators<Kokkos::View<double *, Kokkos::Experimental::SYCL>, Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>::Functor_TestHalfOperators' requested here
  Functor_TestHalfOperators<ViewType, half_type> f_device(h_lhs, h_rhs);
                                                 ^
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:1052:5: note: in instantiation of function template specialization 'Test::__test_half_operators<Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>' requested here
    __test_half_operators<half_t>(h_lhs + cast_to_half(i + 1),
    ^
/soft/restricted/CNDA/updates/2023.05.15.001/oneapi/compiler/eng-20230614/compiler/linux/bin-llvm/../include/sycl/half_type.hpp:530:3: note: candidate function not viable: no known conversion from 'const Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>' to 'const half' for 2nd argument
  OP(<)
  ^
[...]
```
(but `actual_lhs(LT_H_S)   = h_lhs < static_cast<int>(h_rhs);` still works).

Reply:
I would suggest doing
```diff
diff --git a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
index 0b240c88e..b06787cf1 100644
--- a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
+++ b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
@@ -820,18 +820,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs < tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) < rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<impl_type>(lhs) < rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<(T lhs, floating_point_wrapper rhs) {
-    return lhs < T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs < static_cast<impl_type>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -841,18 +839,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs > tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) > rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) > rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>(T lhs, floating_point_wrapper rhs) {
-    return lhs > T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs > static_cast<float>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -862,18 +858,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs <= tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<=(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) <= rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<=(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) <= rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<=(T lhs, floating_point_wrapper rhs) {
-    return lhs <= T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<=(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs <= static_cast<float>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -883,18 +877,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs >= tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>=(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) >= rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>=(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) >= rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>=(T lhs, floating_point_wrapper rhs) {
-    return lhs >= T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>=(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs >= static_cast<float>(rhs);
   }
 
   // Insertion and extraction operators
```

Reply:
No, https://cppinsights.io/s/2155d1dd shows that we would compare doubles in that case (the half_type is first cast to `float` and then to `double`).

Reply:
@masterleinad: OK, thank you for the detailed input. We currently downcast for all mixed precision expressions unless we are mixing with float or double operands, in which case, we upcast the half-type to float or double. https://godbolt.org/z/vaWhahrb1 shows that when mixing a float with a integral-type in binary arithmetic or comparison operators, this same behavior holds: the compiler implicitly downcasts the `long int lhs` to a float. This PR is intended to address a specific bug when mixing half with single and double precision types but may be getting a bit mixed up with some larger feature requests. It's not clear to me whether the compiler will handle upcasting and downcasting for floating point types and integral types involving half-precision operands without being explicit so I pushed this change:
https://github.com/kokkos/kokkos/blob/630a520aa78329b0d8981a8f47cab5ab54d79ce1/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp#L842-L857.
to maintain the proper upcast / downcast behavior in our LT comparison operator.

What do you think? Also are you OK with addressing mixed precision involving integral types during a separate issue / PR?

Reply:
> [godbolt.org/z/vaWhahrb1](https://godbolt.org/z/vaWhahrb1) shows that when mixing a float with a integral-type in binary arithmetic or comparison operators, this same behavior holds: the compiler implicitly downcasts the long int lhs to a float. [...] It's not clear to me whether the compiler will handle upcasting and downcasting for floating point types and integral types involving half-precision operands without being explicit  [...]

Right, and that's the same behavior https://cppinsights.io/s/1cc93aec(slight variation of https://cppinsights.io/s/2155d1dd) shows: An `int` is cast to `float`, for a `double` input we cast the `half_type` to `double` and for a `float` input we don't need extra casts (apart from the one from `half_type` to `float` that is explicit).
This is consistent with https://godbolt.org/z/7se48K5df resp. https://cppinsights.io/s/6e31f3e4 (a slight variation of your example using `double` instead of `int`): We cast the `float` variable to `double`.

Casting integral types to `half_type` looks OKish but I would rather leave selecting the appropriate casts in the comparison operators to the compiler. (although checking for `!std::is_floating_point_v` introduces a comparison operator for arbitrary types like `std::string` which might not what we want).
Can you explain why you don't trust the compiler to do the "right thing" after converting the `half_type` to `float` (which doesn't lose information)?

Reply:
It's not so much about not trusting the compiler to do the "right thing" as much as it is about ensuring we provide these operations in half-precision rather than relying on single precision arithmetic -- In short, it's not clear to me how the compiler can figure out that half-precision operations should be used without being explicit here.

Reply:
> In short, it's not clear to me how the compiler can figure out that half-precision operations should be used without being explicit here.

Right, we wouldn't use `half-precision` operations if we cast to `float` explicitly and then it's a question about correctness vs. performance. 
I guess it would be best to forward the operation to the actual `half_type` in case that it's not just a storage type.

Reply:
Are you OK with merging this bug-fix and addressing integral types in a follow-on?

Reply:
I think it's a pretty easy fix to do
```C++
KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool> 
operator<(const floating_point_wrapper& lhs, const T& rhs) { 
#ifdef KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
  return lhs.val < rhs;
#else
  return static_cast<float>(lhs) < rhs;
#endif
}
```
to actually use half-precision operations if possible. If you want to restrict this overload to `float` and `double` for now and add testing for `int` in a later pull request, that's fine with me.

Reply:
Are you suggesting we remove the member functions for LT,GT,LE, and GE?

Reply:
Can we? Since we implement both `operator<(T, half_type)` and `operator<half_type, T`, I would expect `operator<(half_type, half_type)` to be ambiguous if not explicitly provided.